data-prepper-plugins/dynamodb-source/src/main/java/org/opensearch/dataprepper/plugins/source/dynamodb/export/DataFileScheduler.java

```diff
@@ -0,0 +1,185 @@
+        runLoader.whenComplete(completeDataLoader(dataFilePartition));
+        numOfWorkers.incrementAndGet();
+    }
+
+    @Override
+    public void run() {
+        LOG.debug("Start running Data file Scheduler");
+
+        while (!Thread.interrupted()) {
+            if (numOfWorkers.get() < MAX_JOB_COUNT) {
+                final Optional<SourcePartition> sourcePartition = coordinator.acquireAvailablePartition(DataFilePartition.PARTITION_TYPE);
+
+                if (sourcePartition.isPresent()) {
+                    DataFilePartition dataFilePartition = (DataFilePartition) sourcePartition.get();
+                    processDataFilePartition(dataFilePartition);
+                }
+            }
+            try {
+                Thread.sleep(DEFAULT_LEASE_INTERVAL_MILLIS);
+            } catch (final InterruptedException e) {
+                LOG.info("InterruptedException occurred");
+                break;
+            }
+
+        }
+        LOG.debug("Data file scheduler is interrupted, Stop all data file loaders...");
+        // Cannot call executor.shutdownNow() here
+        // Otherwise the final checkpoint will fail due to SDK interruption.
+        executor.shutdown();
+        DataFileLoader.stopAll();
+    }
+
+    private TableInfo getTableInfo(String tableArn) {
+        GlobalState tableState = (GlobalState) coordinator.getPartition(tableArn).get();
+        TableInfo tableInfo = new TableInfo(tableArn, TableMetadata.fromMap(tableState.getProgressState().get()));
+        return tableInfo;
+    }
+
+    private String getTableArn(String exportArn) {
+        // e.g. given export arn:arn:aws:dynamodb:us-west-2:123456789012:table/Thread/export/01693291918297-bfeccbea
+        // returns: arn:aws:dynamodb:us-west-2:123456789012:table/Thread
+        return exportArn.substring(0, exportArn.lastIndexOf("/export/"));
+    }
+
+    private String getStreamArn(String exportArn) {
+        String tableArn = getTableArn(exportArn);
+        TableInfo tableInfo = getTableInfo(tableArn);
+
+        if (tableInfo.getMetadata().isStreamRequired()) {
+            return tableInfo.getMetadata().getStreamArn();
+        }
+        return null;
+    }
+
+
+    private BiConsumer completeDataLoader(DataFilePartition dataFilePartition) {
+        return (v, ex) -> {
+            numOfWorkers.decrementAndGet();
+            if (ex == null) {
+                exportFileSuccessCounter.increment();
+                // Update global state
+                updateState(dataFilePartition.getExportArn(), dataFilePartition.getProgressState().get().getLoaded());
+                // After global state is updated, mask the partition as completed.
+                coordinator.completePartition(dataFilePartition);
+
+            } else {
+                // The data loader must have already done one last checkpointing.
+                LOG.debug("Data Loader completed with exception");
+                LOG.error(ex.toString());
```

Comment:
Also, do not pass an arbitrary string in for interpolation. This is better: `LOG.error("{}", ex.toString());`

Reply:
Updated.